src/Domain/CoordinateMaps/ProductMapsTimeDep.tpp

```diff
@@ -0,0 +1,413 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include "Domain/CoordinateMaps/ProductMapsTimeDep.hpp"
+
+#include <array>
+#include <boost/none.hpp>
+#include <boost/optional.hpp>
+#include <cstddef>
+#include <functional>
+#include <pup.h>
+#include <utility>
+
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/CoordinateMaps/CoordinateMapHelpers.hpp"
+#include "Utilities/DereferenceWrapper.hpp"
+#include "Utilities/MakeWithValue.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TypeTraits.hpp"
+
+namespace domain {
+namespace CoordMapsTimeDependent {
+namespace product_detail {
+template <typename T, size_t Size, typename Map1, typename Map2,
+          typename Function, size_t... Is, size_t... Js>
+std::array<tt::remove_cvref_wrap_t<T>, Size> apply_call(
+    const std::array<T, Size>& coords, const Map1& map1, const Map2& map2,
+    const Function func, std::integer_sequence<size_t, Is...> /*meta*/,
+    std::integer_sequence<size_t, Js...> /*meta*/) noexcept {
+  using UnwrappedT = tt::remove_cvref_wrap_t<T>;
+  return {
+      {func(
+           std::array<std::reference_wrapper<const UnwrappedT>, sizeof...(Is)>{
+               {coords[Is]...}},
+           map1)[Is]...,
+       func(
+           std::array<std::reference_wrapper<const UnwrappedT>, sizeof...(Js)>{
+               {coords[Map1::dim + Js]...}},
+           map2)[Js]...}};
+}
+
+template <size_t Size, typename Map1, typename Map2, typename Function,
+          size_t... Is, size_t... Js>
+boost::optional<std::array<double, Size>> apply_inverse(
```

Comment:
So if you used a `DataVector` you need a way to signal that one of the points is not in the range of the map. Since that's rather difficult to deal with we decided it only makes sense to have `double`s be support. Does that make sense?